representation/add-logical-knobs.metta

```diff
@@ -0,0 +1,37 @@
+;;;;;;;; addLogicalKnobs ;;;;;;;;;
+;; Creates and adds logical subtree knobs in to a multimap
+;; Params:
+;;   $exemplar: Reference tree containing the target node.
+;;   (mkNodeId $targetId): ID of the target node in the exemplar.
+;;   $addIfInExemplar: If true, include knobs even if in exemplar.
+;;   $map: a multimap to add the new knobs into
+;; Return: a multimap of ($knobSpec $knob) and the updated tree
+
+;; helper function to the addLogicalKnobs
+;; takes a tuple of knobs and return a tuple of knob and knobSpec pairs
+(: mkPair (-> Expression Expression))
+(= (mkPair $tuple)
+(collapse (let*
+(
+    ($knob (superpose $tuple))
+    ($knobSpec (getKnobSpec $knob))
+)
+($knobSpec $knob)
+)))
+
+;; addLogicalKnobs
+(: addLogicalKnobs (-> (Tree $a) NodeId Bool (MultiMap ($k $v)) Expression))
+(= (addLogicalKnobs $exemplar (mkNodeId $targetId) $addIfInExemplar $map) 
```

Comment:
You don't need to deconstruct the id as `(mkNodeId $targetId)` if you are just going to use it as `(mkNodeId $targetId)` in the code. You can just name it as `$nodeId` and everything should work fine.